data-prepper-plugins/blocking-buffer/src/main/java/org/opensearch/dataprepper/plugins/buffer/blockingbuffer/BlockingBuffer.java

```diff
@@ -104,6 +108,7 @@ public void doWrite(T record, int timeoutInMillis) throws TimeoutException {
             LOG.error("Pipeline [{}] - Buffer is full, interrupted while waiting to write the record", pipelineName, ex);
             throw new TimeoutException("Buffer is full, timed out waiting for a slot");
         }
+        updateBufferUsage(1);
```

Comment:
It seems the reason we have this particular signature is that the `AbstractBuffer` calls `doWriteAll` before updating `recordsInBuffer`. Might it make sense to instead add a method to report the buffer status _after_ that is updated? I'm not sure the best name bug possibly `reportBufferUsage(int recordsInBuffer)`?

e.g. in `AbstractBuffer`

```
final int size = records.size();
doWriteAll(records, timeoutInMillis);
recordsWrittenCounter.increment(size);
recordsInBuffer.addAndGet(size);
reportBufferUsage(recordsInBuffer);
```

This method can also have a default implementation rather than be abstract.